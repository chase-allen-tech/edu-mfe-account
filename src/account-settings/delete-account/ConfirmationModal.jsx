import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  Button, Input, Modal, ValidationFormGroup,
} from '@edx/paragon';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { faExclamationCircle, faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { getConfig } from '@edx/frontend-platform';
import messages from './messages';
import Alert from '../Alert';
import PrintingInstructions from './PrintingInstructions';

export class ConfirmationModal extends Component {
  /**
   * @returns String The message id for a short description of the error, suitable for a header or
   * as the error message under an input field.
   */
  getShortErrorMessageId(reason) {
    switch (reason) {
      case 'empty-password':
        return 'account.settings.delete.account.error.no.password';
      case 'invalid-password':
        return 'account.settings.delete.account.error.invalid.password';
      default:
        return 'account.settings.delete.account.error.unable.to.delete';
    }
  }

  renderError(reason) {
    const { errorType, intl } = this.props;

    if (errorType === null) {
      return null;
    }
    const headerMessageId = this.getShortErrorMessageId(errorType);
    const detailsMessageId = reason === 'empty-password'
      ? null
      : 'account.settings.delete.account.error.unable.to.delete.details';

    return (
      <Alert
        className="alert-danger mt-n2"
        icon={<FontAwesomeIcon className="mr-2" icon={faExclamationCircle} />}
      >
        <h6>{intl.formatMessage(messages[headerMessageId])}</h6>
        {detailsMessageId ? (
          <p className="text-danger">{intl.formatMessage(messages[detailsMessageId])}</p>
        ) : null}
      </Alert>
    );
  }

  render() {
    const {
      status,
      errorType,
      intl,
      onCancel,
      onChange,
      onSubmit,
      password,
    } = this.props;
    const open = ['confirming', 'pending', 'failed'].includes(status);
    const passwordFieldId = 'passwordFieldId';
    const invalidMessage = messages[this.getShortErrorMessageId(errorType)];

    // TODO: We lack a good way of providing custom language for a particular site.  This is a hack
    // to allow edx.org to fulfill its business requirements.
    const deleteAccountModalText2MessageKey = getConfig().SITE_NAME === 'edX'
      ? 'account.settings.delete.account.modal.text.2.edX'
      : 'account.settings.delete.account.modal.text.2';

    return (
      <Modal
        open={open}
        title={intl.formatMessage(messages['account.settings.delete.account.modal.header'])}
        body={(
          <div>
            {this.renderError()}
            <Alert
              className="alert-warning mt-n2"
              icon={<FontAwesomeIcon className="mr-2" icon={faExclamationTriangle} />}
            >
              <h6>
                {intl.formatMessage(
                  messages['account.settings.delete.account.modal.text.1'],
                  { siteName: getConfig().SITE_NAME },
                )}
              </h6>
              <p>
                {intl.formatMessage(
                  messages[deleteAccountModalText2MessageKey],
                  { siteName: getConfig().SITE_NAME },
                )}
              </p>
              <p>
                <PrintingInstructions />
              </p>
            </Alert>
            <ValidationFormGroup
              for={passwordFieldId}
              invalid={errorType !== null}
              invalidMessage={intl.formatMessage(invalidMessage)}
            >
              <label className="d-block" htmlFor={passwordFieldId}>
                {intl.formatMessage(messages['account.settings.delete.account.modal.enter.password'])}
              </label>
              <Input
                name="password"
                id={passwordFieldId}
                type="password"
                value={password}
                onChange={onChange}
              />
            </ValidationFormGroup>
          </div>
        )}
        buttons={[
          <Button variant="danger" onClick={onSubmit}>
            {intl.formatMessage(messages['account.settings.delete.account.modal.confirm.delete'])}
          </Button>,
        ]}
        closeText={intl.formatMessage(messages['account.settings.delete.account.modal.confirm.cancel'])}
        renderHeaderCloseButton={false}
        onClose={onCancel}
      />
    );
  }
}

ConfirmationModal.propTypes = {
  status: PropTypes.oneOf(['confirming', 'pending', 'deleted', 'failed']),
  errorType: PropTypes.oneOf(['empty-password', 'server']),
  intl: intlShape.isRequired,
  onCancel: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  password: PropTypes.string.isRequired,
};

ConfirmationModal.defaultProps = {
  status: null,
  errorType: null,
};

export default injectIntl(ConfirmationModal);
