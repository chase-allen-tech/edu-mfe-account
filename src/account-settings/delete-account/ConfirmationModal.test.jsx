import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { IntlProvider, injectIntl } from '@edx/frontend-platform/i18n';

// Modal creates a portal.  Overriding ReactDOM.createPortal allows portals to be tested in jest.
ReactDOM.createPortal = node => node;

import { ConfirmationModal } from './ConfirmationModal'; // eslint-disable-line import/first

const IntlConfirmationModal = injectIntl(ConfirmationModal);

describe('ConfirmationModal', () => {
  let props = {};

  beforeEach(() => {
    props = {
      onCancel: jest.fn(),
      onChange: jest.fn(),
      onSubmit: jest.fn(),
      status: null,
      errorType: null,
      password: 'fluffy bunnies',
    };
  });

  it('should match default closed confirmation modal snapshot', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <IntlConfirmationModal
            {...props}
          />
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match open confirmation modal snapshot', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <IntlConfirmationModal
            {...props}
            status="pending" // This will cause 'modal-backdrop' and 'show' to appear on the modal as CSS classes.
          />
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match empty password confirmation modal snapshot', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <IntlConfirmationModal
            {...props}
            errorType="empty-password"
            status="pending" // This will cause 'modal-backdrop' and 'show' to appear on the modal as CSS classes.
          />
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
