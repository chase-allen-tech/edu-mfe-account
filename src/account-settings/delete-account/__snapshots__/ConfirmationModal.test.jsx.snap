// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationModal should match default closed confirmation modal snapshot 1`] = `
Array [
  <div
    className="fade"
    role="presentation"
  />,
  <div
    className="modal fade"
    role="presentation"
  >
    <div
      aria-labelledby="id1"
      aria-modal={true}
      className=""
      role="dialog"
    >
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={-1}
      />
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={-1}
      />
      <div
        data-focus-lock-disabled="disabled"
        onBlur={[Function]}
        onFocus={[Function]}
        onScrollCapture={[Function]}
        onTouchMoveCapture={[Function]}
        onWheelCapture={[Function]}
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header"
          >
            <h2
              className="modal-title"
              id="id1"
            >
              Are you sure?
            </h2>
          </div>
          <div
            className="modal-body"
          >
            <div>
              <div
                className="alert d-flex align-items-start alert-warning mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-exclamation-triangle fa-w-18 mr-2"
                    data-icon="exclamation-triangle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    You have selected "Delete My Account". Deletion of your account and personal data is permanent and cannot be undone. localhost will not be able to recover your account or the data that is deleted.
                  </h6>
                  <p>
                    If you proceed, you will be unable to use this account to take courses on localhost.
                  </p>
                  <p>
                    <span>
                      You may also lose access to verified certificates and other program credentials. You can make a copy of these for your records before proceeding with deletion.
                    </span>
                  </p>
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  className="d-block"
                  htmlFor="passwordFieldId"
                >
                  If you still wish to continue and delete your account, please enter your account password:
                </label>
                <input
                  aria-describedby=""
                  className="form-control"
                  id="passwordFieldId"
                  name="password"
                  onChange={[MockFunction]}
                  type="password"
                  value="fluffy bunnies"
                />
                <strong
                  className="invalid-feedback"
                  id="passwordFieldId-invalid-feedback"
                >
                  Unable to delete account
                </strong>
              </div>
            </div>
          </div>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-link"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
            <button
              className="btn btn-danger"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Yes, Delete
            </button>
          </div>
        </div>
      </div>
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={-1}
      />
    </div>
  </div>,
]
`;

exports[`ConfirmationModal should match empty password confirmation modal snapshot 1`] = `
Array [
  <div
    className="modal-backdrop show"
    role="presentation"
  />,
  <div
    className="modal show d-block"
    role="presentation"
  >
    <div
      aria-labelledby="id3"
      aria-modal={true}
      className="modal-dialog"
      role="dialog"
    >
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={0}
      />
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={1}
      />
      <div
        data-focus-lock-disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onScrollCapture={[Function]}
        onTouchMoveCapture={[Function]}
        onTouchStart={[Function]}
        onWheelCapture={[Function]}
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header"
          >
            <h2
              className="modal-title"
              id="id3"
            >
              Are you sure?
            </h2>
          </div>
          <div
            className="modal-body"
          >
            <div>
              <div
                className="alert d-flex align-items-start alert-danger mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-exclamation-circle fa-w-16 mr-2"
                    data-icon="exclamation-circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    A password is required
                  </h6>
                  <p
                    className="text-danger"
                  >
                    Sorry, there was an error trying to process your request. Please try again later.
                  </p>
                </div>
              </div>
              <div
                className="alert d-flex align-items-start alert-warning mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-exclamation-triangle fa-w-18 mr-2"
                    data-icon="exclamation-triangle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    You have selected "Delete My Account". Deletion of your account and personal data is permanent and cannot be undone. localhost will not be able to recover your account or the data that is deleted.
                  </h6>
                  <p>
                    If you proceed, you will be unable to use this account to take courses on localhost.
                  </p>
                  <p>
                    <span>
                      You may also lose access to verified certificates and other program credentials. You can make a copy of these for your records before proceeding with deletion.
                    </span>
                  </p>
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  className="d-block"
                  htmlFor="passwordFieldId"
                >
                  If you still wish to continue and delete your account, please enter your account password:
                </label>
                <input
                  aria-describedby="passwordFieldId-invalid-feedback"
                  className="form-control is-invalid"
                  id="passwordFieldId"
                  name="password"
                  onChange={[MockFunction]}
                  type="password"
                  value="fluffy bunnies"
                />
                <strong
                  className="invalid-feedback"
                  id="passwordFieldId-invalid-feedback"
                >
                  A password is required
                </strong>
              </div>
            </div>
          </div>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-link"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
            <button
              className="btn btn-danger"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Yes, Delete
            </button>
          </div>
        </div>
      </div>
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={0}
      />
    </div>
  </div>,
]
`;

exports[`ConfirmationModal should match open confirmation modal snapshot 1`] = `
Array [
  <div
    className="modal-backdrop show"
    role="presentation"
  />,
  <div
    className="modal show d-block"
    role="presentation"
  >
    <div
      aria-labelledby="id2"
      aria-modal={true}
      className="modal-dialog"
      role="dialog"
    >
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={0}
      />
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={1}
      />
      <div
        data-focus-lock-disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onScrollCapture={[Function]}
        onTouchMoveCapture={[Function]}
        onTouchStart={[Function]}
        onWheelCapture={[Function]}
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header"
          >
            <h2
              className="modal-title"
              id="id2"
            >
              Are you sure?
            </h2>
          </div>
          <div
            className="modal-body"
          >
            <div>
              <div
                className="alert d-flex align-items-start alert-warning mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-exclamation-triangle fa-w-18 mr-2"
                    data-icon="exclamation-triangle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    You have selected "Delete My Account". Deletion of your account and personal data is permanent and cannot be undone. localhost will not be able to recover your account or the data that is deleted.
                  </h6>
                  <p>
                    If you proceed, you will be unable to use this account to take courses on localhost.
                  </p>
                  <p>
                    <span>
                      You may also lose access to verified certificates and other program credentials. You can make a copy of these for your records before proceeding with deletion.
                    </span>
                  </p>
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  className="d-block"
                  htmlFor="passwordFieldId"
                >
                  If you still wish to continue and delete your account, please enter your account password:
                </label>
                <input
                  aria-describedby=""
                  className="form-control"
                  id="passwordFieldId"
                  name="password"
                  onChange={[MockFunction]}
                  type="password"
                  value="fluffy bunnies"
                />
                <strong
                  className="invalid-feedback"
                  id="passwordFieldId-invalid-feedback"
                >
                  Unable to delete account
                </strong>
              </div>
            </div>
          </div>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-link"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
            <button
              className="btn btn-danger"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Yes, Delete
            </button>
          </div>
        </div>
      </div>
      <div
        data-focus-guard={true}
        style={
          Object {
            "height": "0px",
            "left": "1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "fixed",
            "top": "1px",
            "width": "1px",
          }
        }
        tabIndex={0}
      />
    </div>
  </div>,
]
`;
